src/Followable.php

```diff
@@ -21,32 +21,93 @@
  */
 trait Followable
 {
+    /**
+     * @return bool
+     */
+    public function isPrivateUser()
+    {
+        return false;
+    }
+
     /**
      * @param \Illuminate\Database\Eloquent\Model|int $user
+     *
+     * @return array
      */
     public function follow($user)
     {
-        $this->followings()->attach($user);
+        $isPending = $user->isPrivateUser() ?: false;
+
+        $this->followings()->attach($user, [
+            'accepted_at' => $isPending ? null : now()
+        ]);
+
+        return ['pending' => $isPending];
     }
 
     /**
      * @param \Illuminate\Database\Eloquent\Model|int $user
-     *
-     * @return int
      */
     public function unfollow($user)
     {
-        return $this->followings()->detach($user);
+        $this->followings()->detach($user);
     }
 
     /**
      * @param \Illuminate\Database\Eloquent\Model|int $user
      *
-     * @return array|array[]
      */
     public function toggleFollow($user)
     {
-        return $this->followings()->toggle($user);
+        $this->isFollowing($user) ? $this->unfollow($user) : $this->follow($user);
+    }
+
+    /**
+     * @param \Illuminate\Database\Eloquent\Model|int $user
+     */
+    public function rejectFollowRequestFrom($user)
+    {
+        if ($user instanceof Model) {
+            $user = $user->getKey();
+        }
+
+        $this->followers()->detach($user);
+    }
+
+    /**
+     * @param \Illuminate\Database\Eloquent\Model|int $user
+     */
+    public function acceptFollowRequestFrom($user)
+    {
+        if ($user instanceof Model) {
+            $user = $user->getKey();
+        }
+
+        $this->followers()->updateExistingPivot($user, ['accepted_at' => now()]);
+    }
+
+    /**
+     * @param \Illuminate\Database\Eloquent\Model|int $user
+     */
+    public function hasRequestedToFollow(Model $user): bool
+    {
+        if ($user instanceof Model) {
+            $user = $user->getKey();
+        }
+
+        /* @var \Illuminate\Database\Eloquent\Model $this */
+        if ($this->relationLoaded('followings')) {
+            return $this->followings
+                ->filter(function ($following) {
+                    return $following->pivot->accepted_at === null;
+                })
+                ->contains($user);
+        }
+
+        return $this->followings()
+            ->whereNull('accepted_at')
```

Comment:
maybe `->wherePivot('accepted_at', null)` better.

Reply:
True, is more explicit